src/parser.c

```diff
@@ -62,18 +62,284 @@ int get_size(var_t *var, type_t *type)
+    else if (lex_accept(T_asterisk))
+        op = OP_mul;
+    else if (lex_accept(T_divide))
+        op = OP_div;
+    else if (lex_accept(T_mod))
+        op = OP_mod;
+    else if (lex_accept(T_lshift))
+        op = OP_lshift;
+    else if (lex_accept(T_rshift))
+        op = OP_rshift;
+    else if (lex_accept(T_log_and))
+        op = OP_log_and;
+    else if (lex_accept(T_log_or))
+        op = OP_log_or;
+    else if (lex_accept(T_eq))
+        op = OP_eq;
+    else if (lex_accept(T_noteq))
+        op = OP_neq;
+    else if (lex_accept(T_lt))
+        op = OP_lt;
+    else if (lex_accept(T_le))
+        op = OP_leq;
+    else if (lex_accept(T_gt))
+        op = OP_gt;
+    else if (lex_accept(T_ge))
+        op = OP_geq;
+    else if (lex_accept(T_ampersand))
+        op = OP_bit_and;
+    else if (lex_accept(T_bit_or))
+        op = OP_bit_or;
+    else if (lex_accept(T_bit_xor))
+        op = OP_bit_xor;
+    else if (lex_peek(T_question, NULL))
+        op = OP_ternary;
+    return op;
+}
+
+int read_numeric_constant(char buffer[])
+{
+    int i = 0;
+    int value = 0;
+    while (buffer[i]) {
+        if (i == 1 && (buffer[i] == 'x')) { /* hexadecimal */
+            value = 0;
+            i = 2;
+            while (buffer[i]) {
+                char c = buffer[i++];
+                value = value << 4;
+                if (is_digit(c))
+                    value += c - '0';
+                c |= 32; /* convert to lower case */
+                if (c >= 'a' && c <= 'f')
+                    value += (c - 'a') + 10;
+            }
+            return value;
         }
-    } while (read_char(0) != '\n');
+        value = value * 10 + buffer[i++] - '0';
+    }
+    return value;
+}
+
+int read_constant_expr_operand()
+{
+    char buffer[MAX_ID_LEN];
+    int value;
+
+    if (lex_peek(T_numeric, buffer)) {
+        lex_expect(T_numeric);
+        return read_numeric_constant(buffer);
+    }
+
+    if (lex_accept(T_open_bracket)) {
+        value = read_constant_expr_operand();
+        lex_expect(T_close_bracket);
+        return value;
+    }
+
+    if (lex_peek(T_identifier, buffer) && !strcmp(buffer, "defined")) {
+        char lookup_alias[MAX_TOKEN_LEN];
+
+        lex_expect(T_identifier); /* defined */
+        lex_expect_internal(T_open_bracket, 0);
+        lex_ident(T_identifier, lookup_alias);
+        lex_expect(T_close_bracket);
+
+        return find_alias(lookup_alias) ? 1 : 0;
+    }
+
+    error("Unexpected token while evaluating constant");
+    return -1;
+}
+
+int read_constant_infix_expr(int precedence)
+{
+    int lhs, rhs, current_precedence;
+    opcode_t op;
+
+    /* Evaluate unary expression first */
+    op = get_operator();
+    current_precedence = get_unary_operator_prio(op);
+    if (current_precedence != 0 && current_precedence >= precedence) {
+        lhs = read_constant_infix_expr(current_precedence);
+
+        switch (op) {
+        case OP_add:
+            break;
+        case OP_sub: {
```

Comment:
Are the braces `{` `}` in the case statements necessary? Can you remove them?

Reply:
There are unnecessary, I will remove them in later commit.

The patch will be committed in #111.